### DISCUSSION TITLE
Query performance

### DISCUSSION THREAD

**Discussion by metmirr** [geldata/gel#3232] [Show and tell]
Hi, I am building a blockchain indexing infra using edgedb. I have observed poor performance on the graphql side as well as the cli side queries.

**Setup**
System:
- Ubuntu 20.04
- 8GB memory
- 160GB Disk
- DigitalOcean Droplet

Edgedb:
- 1.0.0-rc.2
- method=docker
- instance is running in a docker container

Data size:
The total number of rows: 34_691_184

---

**Schema**
```
  type Block {
    index on (.height);

    required property height -> int64;
    required property hash -> str;
  }
```

---

**Sample queries**
Note: since there is no `explain` yet I measure time with a stopwatch

```graphql
# Estimated Execution Time ~1 minute 8 seconds 
{
  Block(order: {height: {dir: DESC}}, first: 50) {
    hash
    height
  }
}
```

```bash
edgedb> select count(Block);
{34691184}
# Estimated execution time ~38 seconds
```

Also the graphql API is publicly available for anyone wants to try it out: http://chainweb-indexer.siriusql.com:10702/db/edgedb/graphql/explore

**Comment by metmirr** [comment]
Additional to the above it seems filtering performance is good:

```graphql
# Execution ~1 second
{
  Block(filter: {height: {gt: 1000000}}, first: 50) {
    hash
    height
  }
}
```

**Comment by elprans** [comment]
This is expected.  Counting large sets in Postgres [is slow](https://wiki.postgresql.org/wiki/Slow_Counting).

**Reply by metmirr** [reply]
Thanks didn’t know that. What about normal queries ? And also graphql query?

**Reply by elprans** [reply]
How log does `SELECT Block { hash, height } ORDER BY .height DESC LIMIT 50;` take in the CLI?

**Reply by elprans** [reply]
I think I found the problem.  The GraphQL translator emits `.height desc empty last`, and `empty last` seems to prevent the use of the index.

**Reply by elprans** [reply]
Opened #3234

**Reply by metmirr** [reply]
It took ~36 seconds from the cli -> `SELECT Block { hash, height } ORDER BY .height DESC LIMIT 50;`

Does it seem something wrong with indexes?

**Reply by metmirr** [reply]
Filtering is normal but ordering is not.

**Reply by elprans** [reply]
Yeah, it's a bug.  We'll fix this in #3234.

**Reply by metmirr** [reply]
Great! Thank you very much! If you dont' mind when you will release fixes?

**Reply by elprans** [reply]
We just shipped 1.0rc3, so this fix will appear in 1.0rc4 in about two weeks.
